### PR TITLE
Reload the dashboard based on time range and filters changes

### PR DESCRIPTION
**Pull request by bfmatei**
**What is this feature?**

This PR brings reload functionality to dashboards when variables and/or time range changes.

In order for this to work correctly, there are a couple of things that needed to happen:
- pass the time range and filters to dashboards fetching API
- introduce feature toggles for the new query params
- cache the dashboards based on the time range and filters besides the scopes

In order to avoid a page refresh, the dashboard reloading mechanism also reloads only the dashboard, not the entire route.

**Why do we need this feature?**

Dynamic dashboards.

**Who is this feature for?**

-

**Which issue(s) does this PR fix?**:

-

**Special notes for your reviewer:**

For testing purposes, you can just hardcode values for `reloadOnTimeRangeChange` and `reloadOnFiltersChange` in `transformSaveModelToScene.ts`

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

**Review by torkelo — COMMENTED** [review]
this feels so wrong on many levels :( 

Can we pause this a bit, want to discuss / understand the underlying problem.

**Comment by roho2000** [comment]
@bfmatei Hi Bogdan, we also need to be able to trigger a reload on a change of *any* variable on the dashboard, not just filters and time range as mentioned in the issue title. Can you confirm that this will be possible?

File: public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts
Comment (torkelo) [inline public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts:316]: why do we need to change the cache key? we always check the version of the dashboard cached and the one returned by the API, so as long as they return different versions of the dashboard we only need to check that

File: public/app/features/dashboard/api/dashboard_api.ts
Comment (torkelo) [inline public/app/features/dashboard/api/dashboard_api.ts:47]: this should not be a global function, but should take time range and filters from the scene

Comment (torkelo) [reply public/app/features/dashboard/api/dashboard_api.ts:47]: you can have two dashboards open at the same time, so not relying on global state is very important

File: pkg/services/featuremgmt/toggles_gen.csv
Comment (dprokop) [inline pkg/services/featuremgmt/toggles_gen.csv:181]: Why do we need two toggles for the reload?

File: public/app/features/dashboard-scene/scene/DashboardScopesFacade.ts
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/DashboardScopesFacade.ts:26]: can we leverage a local scene event here instead of a global event (in case there are two dashboards at the same time)

File: public/app/features/dashboard-scene/scene/ReloadOnFiltersChangeBehavior.ts
Comment (dprokop) [inline public/app/features/dashboard-scene/scene/ReloadOnFiltersChangeBehavior.ts:24]: Don't understand this. So when the `reloadOnFiltersChange` is true it will reload on any variable changed, but otherwise on specified filters? The naming here is quite confusiong. Also - don't like `reloadOnFiltersChange` is either string array or boolean. It's tricky to reason about and handle. Do we even need this to be typed as `string[]`?

File: public/app/features/dashboard-scene/serialization/transformSaveModelToScene.ts
Comment (dprokop) [inline public/app/features/dashboard-scene/serialization/transformSaveModelToScene.ts:264]: Think we should unify this under a single variable. Having two toggles controlling this and two behaviors makes the code veryb "scattered". What if we had a single toogle and behavior, and let the "proxy" decide whether or not the dashboard needs to be refreshed?

File: public/app/features/dashboard-scene/scene/ReloadOnFiltersChangeBehavior.ts
Comment (dprokop) [inline public/app/features/dashboard-scene/scene/ReloadOnFiltersChangeBehavior.ts:19]: Assuming there's a chained variables set in the dashboard, this will trigger the reload multiple times, right @torkelo ?

File: public/app/features/dashboard/api/utils.ts
Comment (bfmatei) [reply public/app/features/dashboard/api/utils.ts:39]: It's the same problem as above: for the initial loading there is no scene :(.

File: pkg/services/featuremgmt/toggles_gen.csv
Comment (bfmatei) [reply pkg/services/featuremgmt/toggles_gen.csv:181]: Thought more granular control can help with future development.

File: public/app/features/dashboard-scene/scene/ReloadOnFiltersChangeBehavior.ts
Comment (bfmatei) [reply public/app/features/dashboard-scene/scene/ReloadOnFiltersChangeBehavior.ts:24]: Yeah. There is the need to reload on specific variables or on all variables. Additionally, after the latest discussions there is also the need to reload on specific labels for adhoc filters.

I'll add multiple different metas for those behaviors:
- `reloadOnAnyVariableChange`: `true` / `false` - it'll reload on any change. This will be the primary check.
- `reloadOnVariablesChange`: `string[]` - it'll reload on variables with the names specified in the array. If an adhoc variable is specified, it will reload on any label change for the specific adhoc variable.
- `reloadOnAdhocLabelsChange`: `string[]` - it'll reload on specific labels in adhoc filters, regardless of the name of the adhoc variable. If a variable is specified in `reloadOnVariablesChange` but specific labels will be specified in `reloadOnAdhocLabelsChange`, the mechanism will ignore the latter.

Comment (dprokop) [reply public/app/features/dashboard-scene/scene/ReloadOnFiltersChangeBehavior.ts:24]: See my comment https://raintank-corp.slack.com/archives/C068ZP1HBGQ/p1728898992409929?thread_ts=1727953353.962049&cid=C068ZP1HBGQ, maybe we can move it here to have the full discussion

File: public/app/features/dashboard-scene/serialization/transformSaveModelToScene.ts
Comment (dprokop) [reply public/app/features/dashboard-scene/serialization/transformSaveModelToScene.ts:264]: yeah, maybe separate behaviors are fine. But having 3 toggles to controll similar thing - that should work together - is a bit overkill to me :)

File: public/app/features/dashboard-scene/scene/DashboardScopesFacade.ts
Comment (bfmatei) [reply public/app/features/dashboard-scene/scene/DashboardScopesFacade.ts:26]: I tried to leverage the local scene event but `DashboardScenePageStateManager` is not part of the scene graph. Passing the instance of `DashboardScenePageStateManager` to `transformSaveModelToScene` and to `createDashboardSceneFromDashboardModel` feels wrong as well. Using `getDashboardScenePageStateManager()` in the behaviors directly lead to circular reference between and there are some nasty side effects.

File: pkg/services/featuremgmt/toggles_gen.csv
Comment (bfmatei) [reply pkg/services/featuremgmt/toggles_gen.csv]: Unified under one feature toggle

File: public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts
Comment (bfmatei) [reply public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts]: Removed the additional cache key params.

File: public/app/features/dashboard-scene/scene/ReloadOnFiltersChangeBehavior.ts
Comment (torkelo) [reply public/app/features/dashboard-scene/scene/ReloadOnFiltersChangeBehavior.ts:19]: yes, and if you have the refresh on time range enabled + refresh on variable it will trigger 1 reload for time range + 1 for each variable

File: public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts
Comment (torkelo) [inline public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts:75]: there should be no keepDashboardFromExploreInLocalStorage anymore, can you pull from main?

File: public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts
Comment (torkelo) [inline public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts:184]: can't find what this params is or where it is passed (where is it called with this params?) why is it a separate function arg and not part of LoadDashboardOptions?

File: public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts:73]: use urlSync handler

File: public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts:77]: can remove this if you use timeRange.urlSync.getUrlState

File: public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts:98]: can be simply replaced by variable.urlSync

File: public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts:61]:
```suggestion
   const varSet  sceneGraph.getVariables(this.parent!);
    return varSet.state.variables.some((variable) => varSet.isVariableLoadingOrWaitingToUpdate(variable);
    
```

File: public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts
Comment (bfmatei) [reply public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts]: Done. TY!

File: public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts
Comment (bfmatei) [reply public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts]: Done. TY!

File: public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts
Comment (bfmatei) [reply public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts]: Done. TY!

File: public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts
Comment (bfmatei) [reply public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts]: Pulled latest main.

File: public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts
Comment (bfmatei) [reply public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts]: `params` are the `queryParams` passed for the reload.

I renamed them to `queryParams` and made them part of `LoadDashboardOptions`

File: pkg/services/featuremgmt/toggles_gen.json
Comment (dprokop) [inline pkg/services/featuremgmt/toggles_gen.json:2279]:
```suggestion
        "description": "Enables the passing of scopes, time range and variables to dashboards fetching in Grafana",
```

**Review by dprokop — APPROVED** [review]
@bfmatei  - gave it another try and it works good. Let's wait for 2nd approval and merge it!

**Comment by bfmatei** [comment]
Heyo @torkelo - can you have a look over this? We're aiming at merging it by tomorrow.

File: public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts
Comment (torkelo) [inline public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts:61]: Is there a way to not rely on a global event? Dashboards can be embedded (used as a component on another page)

File: public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/DashboardReloadBehavior.ts:76]: Emit local event with bubbling (ie  this.publishEvent )

File: public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts
Comment (bfmatei) [reply public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts]: Removed the event all together. We can rely on `getDashboardScenePageStateManager()` to actually retrieve the page manager.

The problem was that the state manager was not part of the scene graph. Creating a separate event bus instance just for this purpose sounds a little bit extreme given the current approach.

WDYT?

File: public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts
Comment (torkelo) [inline public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts:47]: I see ` BackendSrvRequest['params'];` in a few places, this to me is pretty unclear. I would use  `Record<string, any>` or UrlQueryMap

File: public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts
Comment (torkelo) [reply public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts]: looks ok

**Review by torkelo — APPROVED** [review]
looks ok, the only thing confused me a bit as it a bit unclear typing is "BackendSrvRequest['params']" to simply refer to a Record<string, any> (Should probably UrlQueryMap)

File: public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts
Comment (bfmatei) [reply public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts]: Totally forgot about `UrlQueryMap`. Thanks! I just pushed an update with the changes.

**Pull request merged**
